packages/core/src/sagas/incomingActivitySaga.js

```diff
@@ -20,38 +22,45 @@ function last(array, predicate) {
 }
 
 export default function* () {
-  yield whileConnected(function* (directLine, userID) {
-    yield observeEach(directLine.activity$, function* (activity) {
-      activity = { ...activity };
-
-      // Patch activity.from.role to make sure its either "bot", "user", or "channel"
-      if (!activity.from) {
-        activity.from = { role: 'channel' };
-      } else if (!activity.from.role) {
-        if (activity.from.id === userID) {
-          activity.from.role = 'user';
-        } else if (activity.from.id) {
-          activity.from.role = 'bot';
-        } else {
-          activity.from.role = 'channel';
-        }
-      }
-
-      yield put(incomingActivity(activity));
-
-      // Update suggested actions
-      const activities = yield select(({ activities }) => activities);
-      const lastMessageActivity = last(activities, ({ type }) => type === 'message');
-
-      if (
-        lastMessageActivity
-        && lastMessageActivity.from
-        && lastMessageActivity.from.role === 'bot'
-      ) {
-        const { suggestedActions: { actions } = {} } = lastMessageActivity;
-
-        yield put(setSuggestedActions(actions));
-      }
-    });
+  yield whileConnected(observeActivity);
+}
+
+function* observeActivity(directLine, userID) {
+  yield observeEach(directLine.activity$, function* (activity) {
+    activity = patchActivityWithFromRole(activity, userID);
```

Comment:
this suggests that all activities stored in state does have a `from` property. Is this is true why is the condition on L41 needed?

Reply:
Oh yea, that one can be omitted. Fixed.